pkg/scheduler/framework/plugins/noderesources/balanced_allocation.go

```diff
@@ -67,48 +70,82 @@ func (ba *BalancedAllocation) ScoreExtensions() framework.ScoreExtensions {
 }
 
 // NewBalancedAllocation initializes a new plugin and returns it.
-func NewBalancedAllocation(_ runtime.Object, h framework.Handle) (framework.Plugin, error) {
+func NewBalancedAllocation(baArgs runtime.Object, h framework.Handle) (framework.Plugin, error) {
+	args, ok := baArgs.(*config.NodeResourcesBalancedAllocationArgs)
+	if !ok {
+		return nil, fmt.Errorf("want args to be of type NodeResourcesBalancedAllocationArgs, got %T", baArgs)
+	}
+
+	if err := validation.ValidateNodeResourcesBalancedAllocationArgs(nil, args); err != nil {
+		return nil, err
+	}
+
+	resToWeightMap := make(resourceToWeightMap)
+	for _, resource := range (*args).Resources {
+		resToWeightMap[v1.ResourceName(resource.Name)] = resource.Weight
+	}
+
 	return &BalancedAllocation{
 		handle: h,
 		resourceAllocationScorer: resourceAllocationScorer{
 			BalancedAllocationName,
 			balancedResourceScorer,
-			defaultRequestedRatioResources,
+			resToWeightMap,
 		},
 	}, nil
 }
 
-// todo: use resource weights in the scorer function
 func balancedResourceScorer(requested, allocable resourceToValueMap, includeVolumes bool, requestedVolumes int, allocatableVolumes int) int64 {
-	cpuFraction := fractionOfCapacity(requested[v1.ResourceCPU], allocable[v1.ResourceCPU])
-	memoryFraction := fractionOfCapacity(requested[v1.ResourceMemory], allocable[v1.ResourceMemory])
-	// This to find a node which has most balanced CPU, memory and volume usage.
-	if cpuFraction >= 1 || memoryFraction >= 1 {
-		// if requested >= capacity, the corresponding host should never be preferred.
-		return 0
+	resourceToFraction := make(map[v1.ResourceName]float64)
+	var totalFraction float64
+	for name, value := range requested {
+		if _, ok := allocable[name]; ok {
+			fraction := fractionOfCapacity(value, allocable[name])
+			if fraction >= 1 {
+				// if requested >= capacity, the corresponding host should never be preferred.
+				return 0
+			}
+			totalFraction = totalFraction + fraction
+			resourceToFraction[name] = fraction
+		} else {
+			klog.InfoS("requested resource not found for node score calculation", "resource", name)
+			return 0
+		}
 	}
 
 	if includeVolumes && utilfeature.DefaultFeatureGate.Enabled(features.BalanceAttachedNodeVolumes) && allocatableVolumes > 0 {
```

Comment:
uhmm... we should read this feature gate in the defaulting and add volumes to the list of resources.

Reply:
the prerequisite is that pod has any volume asked in the spec. we don't know that when defaults are set, can we?

Reply:
yes, if I understand correctly, 
- read the feature gate in the defaults.go,  adding volumes to the list of default resources if true.
- when calculating `allocatable` and `requested` for the `volume` resource defined in the defaulting resource list, we need to check whether the  `len(pod.Spec.Volumes) > 0` and set the `allocatable` and `requested` accordingly, or else just skip the volume.

Firstly, this is doable, if this is what you meant,  I will address this in a separate review as it is drifting from the intention of this PR, is that okay?

Reply:
Not really drifting. The important question is: what happens if the feature gate is enabled and the user configures the resources in the plugin which don't include volumes?

Reply:
volume is a special case that is handled separately, whether volume should be considered depends on 
- feature gate is enabled or not
- pod requires any volume or not

volume will be considered only when both of above are true , this is why we have a method looks like this,
```
balancedResourceScorer(requested, allocable resourceToValueMap, includeVolumes bool, requestedVolumes int, allocatableVolumes int)
```

And `calculateResourceAllocatableRequest` doesn't consider volume as well, so for now, I think we can keep it as is and handle the volume separately, that's being said, volume is not supposed to be configured in the plugin, configure it in the plugin will return (0, 0).

https://github.com/kubernetes/kubernetes/blob/254c77bb8905b80981fa6e4dc5c98c7616f567c0/pkg/scheduler/framework/plugins/noderesources/resource_allocation.go#L107-L113


Thus, for a pod which do ask for volume, 

- If nothing is configured in the plugin, "cpu", "memory" is considered as defaults, and "volume" is also considered in the runtime.
- If volume is configured in the plugin, volume is skipped for the first pass but get counted in the second pass.

As I said, we can give some refactoring here to move the volume as a default resource when feature gate is enabled, and decide whether count volume in the runtime, this will impact all those node resource plugins not just balance plugin, I will create an issue to track this.

Reply:
pls see: https://github.com/kubernetes/kubernetes/issues/102166

Reply:
This cannot address your concern, for example,
- "cpu" is configured only, but both "cpu" and "volume" are taken into consideration.

let's try to refactoring the code.

Reply:
stuff around volume has been removed, so solve this.